### ISSUE TITLE
[BUG] The pydantic model that has declared an alias cannot correctly receive query parameters

### ISSUE THREAD

**Issue by insistence** [fastapi/fastapi#12402]
### Privileged issue

- [X] I'm @tiangolo or he asked me directly to create an issue here.

### Issue Content

Thank you for all the work you have done. I have initiated the [discussion ](https://github.com/fastapi/fastapi/discussions/12401)as requested, but I think this issue is quite important. Initiating this issue is just to prevent the discussion from being drowned out, and I apologize for any offense.

### Example Code

```python
import uvicorn
from typing import Literal

from fastapi import FastAPI, Query
from pydantic import BaseModel, ConfigDict, Field
from pydantic.alias_generators import to_camel

app = FastAPI()


class FilterParams(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel)

    limit: int = Field(100, gt=0, le=100)
    offset: int = Field(0, ge=0)
    order_by: Literal['created_at', 'updated_at'] = 'created_at'
    tags: list[str] = []


@app.get('/items/')
async def read_items(filter_query: FilterParams = Query()):
    return filter_query


if __name__ == '__main__':
    uvicorn.run(app='app:app')
```

### Description

Running the code in the example above, I encountered an incorrect result when accessing http://127.0.0.1:8000/items/?offset=1&orderBy=updated_at in the browser, orderBy did not receive successfully.
```
{
    "limit": 100,
    "offset": 1,
    "orderBy": "created_at",
    "tags": []
}
```
The correct result should be as follows
```
{
    "limit": 100,
    "offset": 1,
    "orderBy": "updated_at",
    "tags": []
}
```

### Operating System

Windows

### FastAPI Version

0.115.0

### Pydantic Version

2.9.2

### Python Version

3.9.19

**Comment by aless10** [comment]
Did you try with the `Annotated` version as in the docs?
```
from fastapi import FastAPI, Query
from pydantic import BaseModel, Field
from typing_extensions import Annotated, Literal

app = FastAPI()


class FilterParams(BaseModel):
    limit: int = Field(100, gt=0, le=100)
    offset: int = Field(0, ge=0)
    order_by: Literal["created_at", "updated_at"] = "created_at"
    tags: list[str] = []


@app.get("/items/")
async def read_items(filter_query: Annotated[FilterParams, Query()]):
    return filter_query
```

**Comment by Usernamelesss** [comment]
Hello everyone! Can you confirm that it is not possible to have other query params along with the pydantic model ?
```
from typing import Annotated, Literal

import uvicorn
from fastapi import FastAPI, Query
from pydantic import BaseModel, Field

app = FastAPI()


class FilterParams(BaseModel):
    limit: int = Field(100, gt=0, le=100)
    offset: int = Field(0, ge=0)
    order_by: Literal["created_at", "updated_at"] = "created_at"
    tags: list[str] = []


@app.get("/items/")
async def read_items(
        filter_query: Annotated[FilterParams, Query()],
        another_param: Annotated[int, Query()]
):
    print(filter_query, another_param)
```

The generated swagger looks like a bit messy in this case 
![image](https://github.com/user-attachments/assets/22127fd4-cc7a-4dd0-807c-0c47a6416b9c)

**Comment by mjurenka** [comment]
> Hello everyone! Can you confirm that it is not possible to have other query params along with the pydantic model ?
> 
> ```
> from typing import Annotated, Literal
> 
> import uvicorn
> from fastapi import FastAPI, Query
> from pydantic import BaseModel, Field
> 
> app = FastAPI()
> 
> 
> class FilterParams(BaseModel):
>     limit: int = Field(100, gt=0, le=100)
>     offset: int = Field(0, ge=0)
>     order_by: Literal["created_at", "updated_at"] = "created_at"
>     tags: list[str] = []
> 
> 
> @app.get("/items/")
> async def read_items(
>         filter_query: Annotated[FilterParams, Query()],
>         another_param: Annotated[int, Query()]
> ):
>     print(filter_query, another_param)
> ```
> 
> The generated swagger looks like a bit messy in this case ![image](https://private-user-images.githubusercontent.com/40054825/374874000-22127fd4-cc7a-4dd0-807c-0c47a6416b9c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkwMDc4NTEsIm5iZiI6MTcyOTAwNzU1MSwicGF0aCI6Ii80MDA1NDgyNS8zNzQ4NzQwMDAtMjIxMjdmZDQtY2M3YS00ZGQwLTgwN2MtMGM0N2E2NDE2YjljLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDE1VDE1NTIzMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNjZTllMjg2NmNmODlmMjc2YmUzZTc1Y2E0MWJmMjNmOTQwYmRjZWE3YjEwOTE0MzViYjkxMGJmZTc0MjVkY2YmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.a8ylWR2Yxs7BmYMckXcNlBRnNlhgAjFIqssvsHtS4Gs)

Yes I believe you cannot use both Pydantic Models and regular Query parameters. You have to pick one.

**Comment by jd-solanki** [comment]
Refs: https://github.com/fastapi/fastapi/discussions/12379 https://github.com/fastapi/fastapi/discussions/12236 https://github.com/fastapi/fastapi/discussions/12212

**Comment by KristofMerey** [comment]
> Did you try with the `Annotated` version as in the docs?
> ```
> from fastapi import FastAPI, Query
> from pydantic import BaseModel, Field
> from typing_extensions import Annotated, Literal
> 
> app = FastAPI()
> 
> 
> class FilterParams(BaseModel):
>     limit: int = Field(100, gt=0, le=100)
>     offset: int = Field(0, ge=0)
>     order_by: Literal["created_at", "updated_at"] = "created_at"
>     tags: list[str] = []
> 
> 
> @app.get("/items/")
> async def read_items(filter_query: Annotated[FilterParams, Query()]):
>     return filter_query
> ```

Hello, Same issue here. I also tried the annotated version, there is no difference. (And I used a simple alias like `Field(None, alias='test')` instead of an alias generator.)

**Comment by Kyle-sandeman-mrdfood** [comment]
Hi, given what I see in the docs screenshot above, this is the same issue I am facing with Header parameter models.
fastapi==0.115.2

To be clear this does NOT happen with the example in the docs, which is likely the only case that was tested (I don't intend to be rude, this is just an observation)

Given this code:

```python
class CommonHeaders(BaseModel):
    host: str
class CommonHeaders2(BaseModel):
    host2: str

@app.get("/items/")
async def read_items(headers: Annotated[CommonHeaders, Header()],):
                     #headers2: Annotated[CommonHeaders2, Header()]):
    return {}
```
We get image 1: 
![image](https://github.com/user-attachments/assets/05a56fac-0bea-4168-853e-1d9af1aeb732)

When we uncomment the headers2 we get image 2:
![image](https://github.com/user-attachments/assets/caf8534f-a8e2-4e04-9f58-deb4689217f4)

**Comment by KristofMerey** [comment]
Although the origin of these issues can be the same, there are two problems here. This issue is about the alias handling in query parameter models. @Usernamelesss introduced another problem (using multiple query parameter models together), and @Kyle-sandeman-mrdfood 's comment with header parameter models seems very similar to that. @jd-solanki linked question #12212 about this topic. I think, a proper issue should be converted from #12212 as well.

**Comment by RB387** [comment]
Most likely related to this PR:
https://github.com/fastapi/fastapi/pull/12504

**Comment by JoseKilo** [comment]
I'm having the same issue on `fastapi==0.115.6`

As a workaround, I've been able to combine a pydantic model + extra query params by abusing `fastapi.Depends`:

```python
from typing import Annotated

import fastapi
import pydantic

app = fastapi.FastAPI()


class FilterParams(pydantic.BaseModel):
    limit: int
    offset: int


@app.get("/items/")
async def read_items(
    filter_query: Annotated[FilterParams, fastapi.Depends(FilterParams)],
    serial_number: Annotated[str, fastapi.Query(alias="serial-number")],
):
    return {}
```

I.e we can aggregate several query parameters using `Depends` + some callable. In this case, the callable happens to be a pydantic model constructor.
https://fastapi.tiangolo.com/tutorial/dependencies/#create-a-dependency-or-dependable

This seems to handle validation errors as expected.

The only remaining issue is that aliases are still not properly handled if they're part of the pydantic model.
But at least we can have those as independent arguments, which is not possible when annotating the model with `Query()`.

Interestingly, this also works:
```python
class FilterParams(pydantic.BaseModel):
    serial_number: str = pydantic.Field(alias="alias")


@app.get("/items/")
async def read_items(
    filter_query: Annotated[FilterParams, fastapi.Depends(FilterParams)],
```

But it doesn't work if the alias is not a valid python identifier, like `serial-number` in my first example (because of the hyphen `-`).

why ?: because pydantic changes the callable signature if the alias is a valid python identifier, but it doesn't otherwise:
```python
import inspect

class FilterParams(pydantic.BaseModel):
    serial_number: str = pydantic.Field(alias="alias")

inspect.signature(FilterParams)
> <Signature (*, alias: str) -> None>

class FilterParams(pydantic.BaseModel):
    serial_number: str = pydantic.Field(alias="serial-number")

inspect.signature(FilterParams)
> <Signature (*, serial_number: str) -> None>
```
So when `fastapi.Depends` inspects the callable, it doesn't see the alias if it isn't a valid python identifier.

**Comment by Viicos** [comment]
Also reported in https://github.com/pydantic/pydantic/issues/11166:

```python
from fastapi import FastAPI, Query
from pydantic import BaseModel, Field


class ParamFilter(BaseModel):
    model_config = {'extra': 'forbid'}

    age: int = Field(default=30, alias='ag')

app = FastAPI()

@app.get('/hello')
def hello(filter_query: Annotated[ParamFilter, Query()]):
    return {'query': filter_query}
```

This will fail if you try hitting `/hello` without any query params:

```json
{"detail":[{"type":"extra_forbidden","loc":["query","age"],"msg":"Extra inputs are not permitted","input":30}]}
```

In `fastapi.dependencies.utils.request_params_to_args()`, `_get_multidict_value()` is called and this function does not account for aliases when defaulting to the default value:

https://github.com/fastapi/fastapi/blob/dd0b335052619ab5f28749d720aa00a061d6dcce/fastapi/dependencies/utils.py#L716-L737

I'll note that this is somewhat inefficient, FastAPI is not relying on Pydantic validation here (that would account for the default value anyway) and is instead manually fetching the default value. But there's probably design constraints behind this choice.

**Comment by RicarTang** [comment]
Why haven't the changes been committed yet😂

**Comment by JobyGeJo** [comment]
🚫 Query() with a BaseModel: Not Allowed

```python
@app.get("/items/")
async def read_items(filter_query: FilterParams = Query()):
    ...
```

- Query() is used to declare individual query parameters.
- You can’t use Query() to inject an entire Pydantic model.
- This will raise an error: FastAPI expects a plain type (like str, int, etc.), not a Pydantic model.

---

✅ Depends() with a BaseModel: The Correct Way

```python
@app.get("/items/")
async def read_items(filter_query: FilterParams = Depends()):
    ...
```

- Depends() is FastAPI’s way of injecting dependencies.
- FastAPI reads the query parameters and populates the Pydantic model for you, field by field.
- It’s how you cleanly parse groups of query params into structured data.

---

Think of it this way:
- Query(), Path(), Body() — these are field-level declarations.
- Depends() — this is function-level dependency injection, and that’s how FastAPI knows to “build” the whole FilterParams object from query parameters.

⸻

💡 Visual Analogy

```python
# This:
@app.get("/search/")
async def search_items(limit: int = Query(10), offset: int = Query(0)):
    ...

# Is equivalent to:
class SearchParams(BaseModel):
    limit: int = 10
    offset: int = 0
```

Using a model is cleaner when you’re dealing with multiple parameters that logically belong together (like filtering or pagination).

**Comment by jannis701** [comment]
thank you for pointing out the `Depends()`-Syntax to use with pydantic-Models @JobyGeJo . I did not get this possibility from the documentation https://fastapi.tiangolo.com/tutorial/dependencies/ which talks only about functions for `Depends(...)`.

When I use this `Depends()`-syntax, aliases in models are working for me (using `Annotaded` + `Query()` Syntax the paramter `test_paramter` in the following example would not work at all) BUT field_validators are not working anymore... `v` will be an empty object, regardless of `another_test_parameter`'s value (which again works using  `Annotaded` + `Query()` Syntax...)

```python
from fastapi import FastAPI, Query
from pydantic import BaseModel, Field

class FilterParams(BaseModel):
    test_parameter: str = Field(alias='testParameter')
    another_test_parameter: str | None = None

    @field_validator("another_test_parameter", mode="before")
    @classmethod
    def ensure_foobar(cls, v: Any):
        if 'foobar' not in v:
            raise ValueError('"foobar" not found in another_test_parameter')
        return v

app = FastAPI()

# alias does work, field_validator does not work
@app.get("/items/")
async def read_items(filter_query: FilterParams = Depends()):
   return filter_query

# alias does not work, field_validator works
@app.get("/items2/")
async def read_items2(filter_query: Annotated[FilterParams, Query()]):
   return filter_query
```

unfortunately I did not find a way to get both (aliases and field_validator) working together...

**Comment by JobyGeJo** [comment]
Glad it helped @jannis701, and thanks for sharing your findings! Yeah, FastAPI treats Depends() differently because it expects it to load models from the body by default (which explains why lists go to the body, not the query). For query parameters, Annotated[Model, Depends(Query(...))] or other custom handling might be needed to get full control over both aliasing and list parsing.

**Comment by jannis701** [comment]
following the Documentation at https://fastapi.tiangolo.com/tutorial/query-param-models/ I would expect that the `Annotated[FilterParams, Query()]`-Syntax should work, so I would still consider it a bug, that aliases of Pydantic-Fields are broken with this syntax. I found this workaround, but I don't like it, as I would have to include FastApi-Query-Functions in my Pydantic-Models:

```python
from typing import Annotated
from fastapi import FastAPI, Query
from pydantic import BaseModel, Field

class FilterParams(BaseModel):
    test_parameter_pydantic: str = Field(alias='testParameterPydantic') # does not work, I consider this a bug
    test_parameter: str = Query(alias='testParameter') # WORKAROUND: use Query instead of Field
    another_test_parameter: str | None = None
    list_test_parameter: list[str] | None = None

app = FastAPI()

@app.get("/items2/")
async def read_items2(filter_query: Annotated[FilterParams, Query()]):
   return filter_query
```

**Comment by JobyGeJo** [comment]
You're absolutely right in showing that when using `Annotated[FilterParams, Query()]`, the `alias` defined via `Field(...)` is not respected, while it works as expected when using `Depends()`. This reinforces the point that FastAPI handles Pydantic models differently depending on whether you're using `Depends()` or `Query()`.

Also, a small clarification: even with `Depends()`, `list_test_parameter` *is* parsed from the query parameters on `GET` requests — it's not treated as part of the body unless you're using `POST` or explicitly define it as such. So the issue really comes down to the alias support, not the data location in this case.

I agree this feels like inconsistent behavior, and I still consider it a bug in how `Query()` interacts with Pydantic models and field aliases. For now, I’ll stick with using `Depends()` since it preserves alias mappings cleanly and keeps the query parsing behavior intuitive.

**Comment by jannis701** [comment]
It seems to me, that this bug has been fixed here: https://github.com/fastapi/fastapi/commit/88d4f2cb1814392f54011b2bbd3fe55c5f2a3278 but this commit re-introduced it: https://github.com/fastapi/fastapi/commit/55035f440bf852f739e3ccd71b67034016ae9bba#diff-aef3dac481b68359f4edd6974fa3a047cfde595254a4567a560cebc9ccb0673fR765

For me changing [this Line](https://github.com/fastapi/fastapi/blob/master/fastapi/dependencies/utils.py#L783) to     

```python
values[field.alias] = value
```

(as it was done by the first mentioned commit) makes everything work

**Comment by tiangolo** [comment]
It seems this was solved in https://github.com/fastapi/fastapi/pull/14360, available in FastAPI 0.123.3 🎉
